### ISSUE TITLE
Mask eos token for packed dataset

### ISSUE THREAD

**Issue by iankur** [pytorch/torchtune#1177]
Hi,
My current understanding of training with packed dataset is that if we have two sequences packed together, we will compute loss for eos token from the first sequence as input and bos token from the second sequence as target. In the recipes, I can find masking only for the padded tokens towards the end of pack and not for these in between extra tokens. Is this correct?

**Comment by iankur** [comment]
@RdoubleA I was looking into text completion dataset which does not have mask that you mention here
> All of our dataset classes determine labels for loss based on the mask

Otherwise, I agree with what you said and was hoping that will be the case for all datasets.

**Comment by RdoubleA** [comment]
For text completion / continued pre-training, there's no concept of "user" and "assistant" or "input" and "output", there is just a document that you want the model to learn the semantics of. So the labels will be a direct copy of the input tokens offset by one so that the model can learn to predict the next token.
```
input: <s> This is  a    cool  fact  about a   cat from    a      textbook about cats. </s>
             |       |     |    |       |        |         |       |      |      |        |             |            |         |
pred:   This is   a cool fact  about   a    cat from a textbook about     cats   </s>
```
None of these tokens should be masked out.

**Comment by iankur** [comment]
My question was about token `</s>` at the end of first sequence and before the start of next sequence when we use packing with text completion dataset. In this case, we will have something like
```
input: <s> This is  a   cool  fact  about a   cat from   a  textbook about cats. </s>  <s> This is ...
        |   |    |  |    |     |     |    |    |   |     |      |      |    |      |    |  
pred:  This is   a cool fact  about  a    cat from a textbook about  cats  </s>   <s> This  ...
```

and there is no way to mask this token or loss on it afterwards with current code. I am not sure what the desired behavior should be but it can affect other tasks such as statistics of activations. Please feel free to close the issue if this not relevant.

**Comment by RdoubleA** [comment]
Ah I see, so if I understand correctly you want to mask out the in between </s> tokens for packed sequences, but not the <s> tokens?

You're right that it can impact activations and model output - but I would still say you want to keep the loss for those tokens. The model needs to learn when a document ends even within a packed sequence so the data should keep those boundaries.
